test/metabase/query_processor_test/model_test.clj

```diff
@@ -0,0 +1,67 @@
+   [metabase.lib.metadata :as lib.metadata]
+   [metabase.lib.metadata.jvm :as lib.metadata.jvm]
+   [metabase.query-processor :as qp]
+   [metabase.test :as mt]))
+
+(deftest ^:parallel model-self-join-test
+  (testing "Field references from model joined a second time can be resolved (#48639)"
+    (let [mp (lib.metadata.jvm/application-database-metadata-provider (mt/id))]
+      (mt/with-temp [:model/Card base-model
+                     {:dataset_query
+                      (-> (lib/query mp (lib.metadata/table mp (mt/id :products)))
+                          (lib/join (-> (lib/join-clause (lib.metadata/table mp (mt/id :reviews))
+                                                         [(lib/=
+                                                           (lib.metadata/field mp (mt/id :products :id))
+                                                           (lib.metadata/field mp (mt/id :reviews :product_id)))])
+                                        (lib/with-join-fields :all)))
+                          lib.convert/->legacy-MBQL)
+                      :database_id (mt/id)
+                      :name "Products+Reviews"
+                      :type :model}
+                     :model/Card consumer-model
+                     {:dataset_query
+                      (as-> (lib/query mp (lib.metadata/card mp (:id base-model))) $q
+                        (lib/aggregate $q (lib/sum (->> $q
+                                                        lib/available-aggregation-operators
+                                                        (m/find-first (comp #{:sum} :short))
+                                                        :columns
+                                                        (m/find-first (comp #{"Price"} :display-name)))))
+                        (lib/breakout $q (-> (m/find-first (comp #{"Reviews → Created At"} :display-name)
+                                                           (lib/breakoutable-columns $q))
+                                             (lib/with-temporal-bucket :month))))
+                      :database_id (mt/id)
+                      :name "Products+Reviews Summary"
+                      :type :model}]
+        (let [question (as-> (lib/query mp (lib.metadata/card mp (:id base-model))) $q
+                         (lib/breakout $q (-> (m/find-first (comp #{"Reviews → Created At"} :display-name)
+                                                            (lib/breakoutable-columns $q))
+                                              (lib/with-temporal-bucket :month)))
+                         (lib/aggregate $q (lib/avg (->> $q
+                                                         lib/available-aggregation-operators
+                                                         (m/find-first (comp #{:avg} :short))
+                                                         :columns
+                                                         (m/find-first (comp #{"Rating"} :display-name)))))
```

Comment:
I really like the way you re-use query that's being built to get the column information here and in other places.

Reply:
It looks ugly, but it simulates the way the query is built in the FE.